src/databricks/labs/ucx/assessment/jobs.py

```diff
@@ -104,3 +120,211 @@ def snapshot(self) -> Iterable[JobInfo]:
+            hash_values.extend(self._retrieve_hash_values_from_condition_task(task.condition_task))
+        if task.run_job_task is not None and task.run_job_task.job_id is not None:
+            hash_values.append(str(task.run_job_task.job_id))
+        if task.git_source is not None:
+            hv = self._retrieve_hash_values_from_git_source(task.git_source)
+            hash_values.extend(hv)
+
+        return [str(value) for value in hash_values if value is not None]
+
+    def _check_task_for_failures(self, task_cluster: compute.ClusterSpec) -> list[str]:
+        # there are several potential cases that we need to check:
+        # 1. Azure SPNs in the configuration
+        # 2. AWS Instance Profiles in the configuration
+        # Any fs-related configuration, e.g. sts or fs.azure.account.key
+        _failures = []
+        if task_cluster.aws_attributes is not None and task_cluster.aws_attributes.instance_profile_arn is not None:
+            _failures.append("AWS instance profile in the configuration")
+        if task_cluster.spark_conf is not None:
+            azure_failures = self.check_spark_conf(task_cluster.spark_conf, "Task cluster")
+            _failures.extend(azure_failures)
+            for key_regexp in self._FS_LEVEL_KEY_REGEXP:
+                for key in task_cluster.spark_conf.keys():
+                    if re.match(key_regexp, key):
+                        _failures.append(f"fs-related configuration: {key}")
+        return _failures
+
+    @staticmethod
+    def _task_needs_compatibility_check(task: RunTask) -> bool:
+        """
+        # we recognize a task as a potentially incompatible one if:
+        # 1. cluster is not configured with data security mode
+        # 2. cluster's DBR version is greater than 11.3
+        """
+        if task.new_cluster is not None and task.new_cluster.data_security_mode is None:
+            dbr_info = parse_spark_version(task.new_cluster.spark_version)
+            if dbr_info.dbr_version:
+                return dbr_info.dbr_version >= DBRVersion(11, 3)
+
+        return False
+
+    def _assess_job_runs(self, submit_runs: Iterable[BaseRun]) -> Iterable[SubmitRunInfo]:
+        """
+        Assessment logic:
+        1. For each submit run, we analyze all tasks inside this run.
+        2. Per each task, we calculate a unique hash based on the _retrieve_hash_values_from_task function
+        3. Then we coalesce all task hashes into a single hash for the submit run
+        4. Coalesce all runs under the same hash into a single pseudo-job
+        5. Return a list of pseudo-jobs with their assessment results
+        """
+        result: dict[str, SubmitRunInfo] = {}
+        runs_per_hash: dict[str, list[int | None]] = {}
+        for submit_run in submit_runs:
+            hashable_items = []
+            all_tasks: list[RunTask] = submit_run.tasks if submit_run.tasks is not None else []
+            failures_per_task: dict[str, list[str]] = {}
+
```

Comment:
refactor to re-use parts of `databricks.labs.ucx.assessment.jobs.JobsCrawler._assess_jobs`